### PR TITLE
P-384/521 fallback to small implementation when OPENSSL_SMALL is set

### PR DESCRIPTION
**Pull request by dkostic**
### Issues:
N/A

### Description of changes: 
We should use the generic `EC_GFp_mont_method` for P-384/521 when OPENSSL_SMALL flag is defined.

### Call-outs:
Point out areas that need special attention or support during the review process. Discuss architecture or design changes.

### Testing:
How is this change tested (unit tests, fuzz tests, etc.)? Are there any testing steps to be verified by the reviewer?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and 
the ISC license.

### CODE REVIEW

**Review by andrewhop — COMMENTED** [review]
What is the size of libcrypto before/after this cahnge? Have you checked that EC_GFp_nistp521_method/EC_GFp_nistp384_method are not in the final library when SMALL is set?

**Review by andrewhop — APPROVED** [review]

File: crypto/ecdh_extra/ecdh_test.cc [hunk 1]
```diff
@@ -130,12 +130,19 @@ TEST(ECDHTest, TestVectors) {
   });
 }
 
-static int has_uint128_and_not_small() {
+// Returns 1 if the curve defined by |nid| is using Montgomery representation
```
Comment (nebeid) [inline crypto/ecdh_extra/ecdh_test.cc:133]: The comment in https://github.com/aws/aws-lc/blob/adea0dbcc26b9523e45dcdbe22ea641747734cc5/crypto/ecdh_extra/ecdh_test.cc#L180-L185 needs to be adjusted:

    // The following call converts the point to Montgomery form for P-256/384.
    // For P-224, when the functions from simple.c are used, i.e. when
    // group->meth = EC_GFp_nistp224_method, the coordinate representation
    // is not changed. This is determined based on compilation flags in ec.c
    // that are duplicated below in is_curve_using_mont_felem_impl().
    // For P-521, the plain non-Motgomery representation is always used also except for 
    // the flag used in the same function.

Or feel free to edit it to make it clearer.

File: crypto/fipsmodule/ec/ec_test.cc [hunk 2]
```diff
@@ -1845,12 +1845,19 @@ static bool HasSuffix(const char *str, const char *suffix) {
   return strcmp(str + str_len - suffix_len, suffix) == 0;
 }
 
-static int has_uint128_and_not_small() {
+// Returns 1 if the curve defined by |nid| is using Montgomery representation
+// for field elements (based on the build configuration). Returns 0 otherwise.
+static int is_curve_using_mont_felem_impl(int nid) {
```
Comment (nebeid) [inline crypto/fipsmodule/ec/ec_test.cc:1850]: Same comment on https://github.com/aws/aws-lc/blob/adea0dbcc26b9523e45dcdbe22ea641747734cc5/crypto/fipsmodule/ec/ec_test.cc#L1891-L1896 as on ecdh_test.cc

File: crypto/fipsmodule/ec/p384.c [hunk 3]
```diff
@@ -30,7 +32,7 @@
 //   #define p384_felem_add(out, in0, in1) bignum_add_p384(out, in0, in1)
 // when s2n-bignum is used.
 //
-#if !defined(OPENSSL_NO_ASM) && \
+#if !defined(OPENSSL_NO_ASM) && !defined(OPENSSL_SMALL) && \
```
Comment (nebeid) [inline crypto/fipsmodule/ec/p384.c:35]: Is this check now a duplicate?

Comment (dkostic) [reply crypto/fipsmodule/ec/p384.c]: yes, removed it.

File: crypto/fipsmodule/ec/ec_test.cc (hunk 2 above)
Comment (dkostic) [reply crypto/fipsmodule/ec/ec_test.cc:1850]: fixed.

File: crypto/fipsmodule/ec/ec_test.cc [hunk 4]
```diff
@@ -1908,12 +1906,10 @@ TEST(ECTest, LargeXCoordinateVectors) {
     OPENSSL_memset(key.get()->pub_key->raw.Z.bytes, 0, len);
     key.get()->pub_key->raw.Z.bytes[0] = 1;
 
-    // As mentioned, for P-224 and P-521, setting the raw point directly
-    // with the coordinates still passes |EC_KEY_check_fips|.
-    // For P-256 and 384, the failure is due to that the coordinates are
-    // not in Montgomery representation, hence the checks fail earlier in
-    // |EC_KEY_check_key| in the point-on-the-curve calculations, which use
-    // Montgomery arithmetic.
+    // |EC_KEY_check_fips| firs calls the |EC_KEY_check_key| function that
```
Comment (nebeid) [inline crypto/fipsmodule/ec/ec_test.cc:1909]:
```suggestion
    // |EC_KEY_check_fips| first calls the |EC_KEY_check_key| function that
```

File: crypto/ecdh_extra/ecdh_test.cc [hunk 5]
```diff
@@ -203,12 +201,11 @@ TEST(ECDHTest, InvalidPubKeyLargeCoord) {
...
     peer_key.get()->pub_key->raw.Z.bytes[0] = 1;
 
-    // As mentioned, for P-224 and P-521, setting the raw point directly
-    // with the coordinates still passes |EC_KEY_check_fips|.
-    // For P-256 and 384, the failure is due to that the coordinates are
-    // not in Montgomery representation, hence the checks fail earlier in
-    // |EC_KEY_check_key| in the point-on-the-curve calculations, which use
-    // Montgomery arithmetic.
+    // |ECDH_compute_key_fips| calls |EC_KEY_check_fips| that calls
+    // |EC_KEY_check_key|function which checks if the computed key point is on
```
Comment (nebeid) [inline crypto/ecdh_extra/ecdh_test.cc:205]:
```suggestion
    // |EC_KEY_check_key| function which checks if the computed key point is on
```

**Review by nebeid — APPROVED** [review]

**Review by andrewhop — APPROVED** [review]

**Pull request merged**